source/logging.rst

```diff
@@ -0,0 +1,309 @@
+per-component level. The below components currently exist and correspond to the
+listed specifications; this list is expected to grow over time.
+
+.. list-table::
+   :header-rows: 1
+   :widths: 1 1 1
+
+   * - Component Name
+     - Specification(s)
+     - Environment Variable
+
+   * - Command
+     - `Command Monitoring <../command-monitoring/command-monitoring.rst>`__
+     - ``MONGODB_LOGGING_COMMAND``
+
+   * - SDAM
+     - `Server Discovery and Monitoring
+       <../server-discovery-and-monitoring/server-discovery-and-monitoring.rst>`__
+     - ``MONGODB_LOGGING_SDAM``
+
+   * - Server Selection
+     - `Server Selection <../server-selection/server-selection.rst>`__
+     - ``MONGODB_LOGGING_SERVER_SELECTION``
+
+   * - Connection
+     - `Connection Monitoring and Pooling
+       <../connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst>`__
+     - ``MONGODB_LOGGING_CONNECTION``
+
+Severity Levels
+---------------
+Driver specifications defining log messages may use any of the following levels,
+inspired by the Syslog Protocol as described in `RFC 5424
+<https://tools.ietf.org/html/rfc5424>`__:
+
+.. list-table::
+   :header-rows: 1
+   :widths: 1 1 1 1
+
+   * - Code
+     - Level Name
+     - Meaning
+     - Environment Variable value (case-insensitive)
+
+   * - N/A
+     - Off
+     - Nothing is logged.
+     - off
+
+   * - 0
+     - Emergency
+     - N/A
+     - emergency
+
+   * - 1
+     - Alert
+     - N/A
+     - alert
+
+   * - 2
+     - Critical
+     - N/A
+     - critical
+
+   * - 3
+     - Error
+     - Any error that we are unable to report to the user via API.
+     - error
+
+   * - 4
+     - Warning
+     - Indicates a situation where undesirable application behavior may occur.
+       Example: The driver ignores an unrecognized option in a URI.
+     - warn
+
+   * - 5
+     - Notice
+     - Indicates an event that is unusual but not problematic. Example: a
+       change stream is automatically resumed.
+     - notice
+
+   * - 6
+     - Informational
+     - High-level information about normal driver behavior. Example:
+       ```MongoClient`` creation or close.
+     - info
+
+   * - 7
+     - Debug
+     - Detailed information that may be helpful when debugging the application.
+       Example: A command starting.
+     - debug
+
+   * - 8
+     - Trace
+     - Very fine-grained details related to logic flow. Example: entering and
+       exiting function bodies.
+     - trace
+
+Note that the Emergency, Alert, and Critical levels have been intentionally left
+undefined. At the time of writing this specification, we do not expect any
+driver specifications to need to log at these levels, but we have included them
+in the list of permitted levels for consistency with Syslog and so that they
+may be used in the future if needed.
+
+Not all logging frameworks will necessarily support all of these levels. If an
+equivalent level is not available, drivers SHOULD emit messages for that level
```

Comment:
I think "MUST" is more appropriate here in favour of "SHOULD", as the rules for dealing with missing severities are quite clear.

At the same time, I'm wondering whether downgrading is the most sensible options, or whether we want error messages (severities "error", "critical", "alert", and "emergency") to be upgraded if they don't exist. For informational messages, it's ok if they get missed but we should err on the side of caution when logging error messages.

Reply:
Using MUST over SHOULD is fine with me.

Re missing messages: from my research into the logging frameworks various drivers have said they would use, I believe the existence of an "error" level is universal.

So I guess the question is really what do we do with "critical", "alert", and "emergency" level messages if their corresponding levels don't exist. I'm fine with mapping them to more severe levels if so, though we could also defer a decision on that until if the time ever comes when we actually need a level higher than "error".

Reply:
Deferring that decision until we have a need for those levels sounds fine to me.

Reply:
I've switched to using MUST. I also added a note in the future work section about defining the unused levels, which mentions the spec should define how to map them to other levels if they are unsupported.